pyro/contrib/gp/models/gplvm.py

```diff
@@ -0,0 +1,131 @@
+        # to a tensor X of shape 150x2, we will use GPLVM.
+        # First, define the initial values for X_loc parameter:
+        >>> X_loc = torch.zeros(150, 2)
+        # Then, define a Gaussian Process model with input X_loc and output y:
+        >>> kernel = gp.kernels.RBF(input_dim=2, lengthscale=torch.ones(2))
+        >>> Xu = torch.zeros(20, 2)  # initial inducing inputs of sparse model
+        >>> gpmodel = gp.models.SparseGPRegression(X_loc, y, kernel, Xu)
+        # Finally, wrap gpmodel by GPLVM, optimize, and get the "learned" mean of X:
+        >>> gplvm = gp.models.GPLVM(gpmodel)
+        >>> gplvm.optimize()
+        >>> X = gplvm.get_param("X_loc")
+
+    Reference:
+
+    [1] Bayesian Gaussian Process Latent Variable Model
+    Michalis K. Titsias, Neil D. Lawrence
+
+    :param ~pyro.contrib.gp.models.model.GPModel base_model: A Pyro Gaussian Process
+        model object. Note that ``base_model.X`` will be the initial value for the
+        variational parameter ``X_loc``.
+    :param str name: Name of this model.
+    """
+    def __init__(self, base_model, name="GPLVM"):
+        super(GPLVM, self).__init__(name)
+        if base_model.X.dim() != 2:
+            raise ValueError("GPLVM model only works with 2D latent X, but got "
+                             "X.dim() = {}.".format(base_model.X.dim()))
+        self.base_model = base_model
+        self.y = self.base_model.y
+
+        self.X_loc = Parameter(self.base_model.X)
+
+        C = self.X_loc.shape[1]
+        X_scale_tril_shape = self.X_loc.shape + (C,)
+        Id = torch.eye(C, out=self.X_loc.new_empty(C, C))
+        X_scale_tril = Id.expand(X_scale_tril_shape)
+        self.X_scale_tril = Parameter(X_scale_tril)
+        self.set_constraint("X_scale_tril", constraints.lower_cholesky)
+
+        self._call_base_model_guide = True
+
+    def model(self):
+        self.set_mode("model", recursive=False)
+
+        # sample X from unit multivariate normal distribution
+        zero_loc = self.X_loc.new_zeros(self.X_loc.shape)
+        C = self.X_loc.shape[1]
+        Id = torch.eye(C, out=self.X_loc.new_empty(C, C))
+        X_name = param_with_module_name(self.name, "X")
+        X = pyro.sample(X_name, dist.MultivariateNormal(zero_loc, scale_tril=Id)
+                                    .independent(zero_loc.dim()-1))
+
+        self.base_model.set_data(X, self.y)
+        self.base_model.model()
+
+    def guide(self):
+        self.set_mode("guide", recursive=False)
+
+        # sample X from variational multivariate normal distribution
+        X_loc = self.get_param("X_loc")
+        X_scale_tril = self.get_param("X_scale_tril")
+        X_name = param_with_module_name(self.name, "X")
+        X = pyro.sample(X_name,
+                        dist.MultivariateNormal(X_loc, scale_tril=X_scale_tril)
+                            .independent(X_loc.dim()-1))
+
+        self.base_model.set_data(X, self.y)
```

Comment:
Is this where you handle the D independent GPs which model the D-dimensions in `y` ?

Reply:
Yes, it depends on which `base_model` we use. For VariationalGP and VariationalSparseGP models, we have an arbitrary independent shape for the GP latent f (as long as its shape can be broadcasted to y.shape). For SparseGPRegression and GPRegression, currently, we just have 1 GP governs all D-dimensions of `y`..

Reply:
Very cool, so a lot of the machinery you need for GPLVM is already in there. This may be the shortest GPLVM implementation ever :)